source/04_best_practice/buffer.rst

```diff
@@ -0,0 +1,172 @@
+        buffer.push('a', meta={"priority": 2.0})
+    buffered_data = buffer.sample(3)
+
+
+**Sample cloning**
+
+By default, for mutable objects stored in a buffer (such as list, np.array, torch.tensor, etc.), the sampling operation in fact returns a reference to that object.
+If the user subsequently makes changes to the content of the reference, it may cause the corresponding content in the sample pool to change as well.
+In some application scenarios, the user may expect the data in the sample pool to remain unchanged, and this can be done by using the \ **clone_object middleware**\  to return a copy of the object in the buffer at sampling time.
+In this way, modifications to the copy contents do not affect the original data in the buffer. \ **Sample cloning significantly increases the sampling elapsed time**\.
+
+.. code-block:: python
+    
+    from ding.data.buffer.middleware import clone_object
+
+    buffer = DequeBuffer(size=10)
+    buffer.use(clone_object())
+
+
+**Group sampling**
+
+In some specific environments or algorithms, users may wish to collect, store, and process samples in entire episodes.
+For example, in chess, Go, or card games where players are only rewarded at the end of the game, algorithms solving such tasks often want to process the entire game, and some algorithms like Hindsight Experience Replay (HER) need to sample complete episodes and process them in episodic units.
+In this case, the user can use group sampling to achieve this goal.
+
+- **Custom implementation via atomic operations**
+
+  When storing samples, you can add "episode" information to the meta to specify the episode to which the sample belongs, and when sampling, you can set groupby="episode" to enable group sampling with the episode keyword. \ **Sampling in groups can seriously increase the sampling time**\.
+
+  .. code-block:: python
+
+      buffer = DequeBuffer(size=10)
+
+      # When storing data, the user needs to add grouping information to the meta, e.g., "episode" as the grouping keyword, and the corresponding value is the specific group
+      buffer.push("a", {"episode": 1})
+      buffer.push("b", {"episode": 2})
+      buffer.push("c", {"episode": 2})
+
+      # Grouping according to the keyword "episode" requires that the number of different groups in the buffer is not less than the number of samples.
+      grouped_data = buffer.sample(2, groupby="episode")
+
+- **Implementation through middleware**
+
+  In the R2D2 algorithm, for example, the training samples are passed through the LSTM network in episodes, where group sampling is required.
+  In DI-engine, each env. corresponds to a separate decision track, so using env. as a keyword for grouping can have the effect of distinguishing between episodes.
+  The relevant code example is as follows, the full version of the code can be seen at \ `./ding/example/r2d2.py <https://github.com/opendilab/DI-engine/blob/main/ding/example/r2d2.py>`_：
```

Comment:
The code for R2D2 using group sampling lies below, the full version can be found at ...